caluma/caluma_form/structure.py

```diff
@@ -144,6 +144,18 @@ def __init__(self, document, form, question=None, parent=None):
         self.question = question
         self._fields = None
         self._sub_forms = None
+        self._main_form = None
+
+    @property
+    def main_form(self):
+        if self._main_form is None:
+            if hasattr(self.document.family, "work_item") and hasattr(
+                self.document.family.work_item.case.family, "document"
+            ):
+                self._main_form = (
+                    self.document.family.work_item.case.family.document.form.slug
+                )
+        return self._main_form
```

Comment:
Two things: One, naming: I think it might be confusing, as the "main form" could also be the the main form of the current document, for example if you're in a row-doc or similar. Suggestion: "main_case_form" or something similar maybe?

Second: If for some reason the inner `if` does not find the main form, it will not cache the result in `self._main_form` and try to do the whole resolution-chain again, possibly triggering quite a few DB queries. I'd suggest caching that case as well

Reply:
Fixed the naming and caching :+1: